repositories.go

```diff
@@ -0,0 +1,234 @@
+	repository := NewRepository(id, repo)
+	p.repositories = append(p.repositories, repository)
+}
+
+// AddGit opens a new git repository and adds it to the pool. It
+// also sets its path as ID.
+func (p *RepositoryPool) AddGit(path string) (string, error) {
+	repo, err := git.PlainOpen(path)
+	if err != nil {
+		return "", err
+	}
+
+	p.Add(path, repo)
+
+	return path, nil
+}
+
+// GetPos retrieves a repository at a given position. It returns false
+// as second return value if the position is out of bounds.
+func (p *RepositoryPool) GetPos(pos int) (*Repository, bool) {
+	if pos >= len(p.repositories) {
+		return nil, false
+	}
+
+	return &p.repositories[pos], true
+}
+
+// RepoIter creates a new Repository iterator
+func (p *RepositoryPool) RepoIter() (*RepositoryIter, error) {
+	iter := &RepositoryIter{
+		pos:  0,
+		pool: p,
+	}
+
+	return iter, nil
+}
+
+// RepositoryIter iterates over all repositories in the pool
+type RepositoryIter struct {
+	pos  int
+	pool *RepositoryPool
+}
+
+// Next retrieves the next Repository. It returns io.EOF as error
+// when there are no more Repositories to retrieve.
+func (i *RepositoryIter) Next() (*Repository, error) {
+	r, ok := i.pool.GetPos(i.pos)
+	if !ok {
+		return nil, io.EOF
+	}
+
+	i.pos++
+
+	return r, nil
+}
+
+// Close finished iterator. It's no-op.
+func (i *RepositoryIter) Close() error {
+	return nil
+}
+
+// RowRepoIterImplementation is the interface needed by each iterator
+// implementation
+type RowRepoIterImplementation interface {
+	NewIterator(*Repository) (RowRepoIterImplementation, error)
+	Next() (sql.Row, error)
+	Close() error
+}
+
+// RowRepoIter is used as the base to iterate over all the repositories
+// in the pool
+type RowRepoIter struct {
+	repositoryIter *RepositoryIter
+	implementation RowRepoIterImplementation
+
+	wg    sync.WaitGroup
+	done  chan bool
+	err   chan error
+	repos chan *Repository
+	rows  chan sql.Row
+}
+
+// NewRowRepoIter initializes a new repository iterator.
+//
+// * pool: is a RepositoryPool we want to iterate
+// * impl: implementation with RowRepoIterImplementation interface
+//     * NewIterator: called when a new repository is about to be iterated,
+//         returns a new RowRepoIterImplementation
+//     * Next: called for each row
+//     * Close: called when a repository finished iterating
+func NewRowRepoIter(pool *RepositoryPool,
+	impl RowRepoIterImplementation) (*RowRepoIter, error) {
+
+	rIter, err := pool.RepoIter()
+	if err != nil {
+		return nil, err
+	}
+
+	rowRepoIter := RowRepoIter{
+		repositoryIter: rIter,
+		implementation: impl,
+		done:           make(chan bool),
+		err:            make(chan error),
+		repos:          make(chan *Repository),
+		rows:           make(chan sql.Row),
+	}
+
+	go rowRepoIter.fillRepoChannel()
+
+	wNum := runtime.NumCPU()
+
+	for i := 0; i < wNum; i++ {
+		rowRepoIter.wg.Add(1)
+
+		go rowRepoIter.rowReader(i)
+	}
+
+	go func() {
+		rowRepoIter.wg.Wait()
```

Comment:
Is it ok having a goroutine only to wait for the waitGroup, if the goroutine where this one is launched ends, it doesn't wait anyway

Reply:
It's waiting until all the goroutines finish to close the rows channel. It can be because all finished its work or there was a problem. I thought it was better to close this in only one place.